### PR TITLE
UPDATE Shell:

### PR DESCRIPTION
**Pull request by riblanc**
Defining a more generic Shell structure, to be able to declare a Shell with a given builtins file and custom reader / writer.
Builtins now take the shell context as first parameter.

This pull request introduces a hook and a config system.

Different hooks are available and can be bound such as:
- ``on_init``: called during Shell.init call.
- ``on_error``: called when an error occurs.
- ``pre_prompt``: called before the shell attempt to read a line.
- ``pre_cmd``: called before calling a builtin.
- ``post_cmd``: called after calling a builtin (only if builtin there isn't any error)

A new "colors" module is added, defining some useful constants and can be import via ``import("colors")``

### CODE REVIEW

File: src/drivers/acpi/acpi.zig [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const ft = @import("../../ft/ft.zig");
 const tty = @import("../../tty/tty.zig");
-const utils = @import("../../shell/utils.zig");
+const utils = @import("../../shell/default/utils.zig");
```
Comment (jgiron42) [inline src/drivers/acpi/acpi.zig:3]: utils is used here for colors in the logger, if the use of these constant has nothing to do with the shell, maybe they should be in another file

File: src/kernel.zig [hunk 2]
```diff
@@ -1,8 +1,13 @@
-const keyboard = @import("./tty/keyboard.zig");
-const printk = @import("./tty/tty.zig").printk;
-const shell = @import("./shell.zig").shell;
+const tty = @import("./tty/tty.zig");
+const DefaultShell = @import("shell/default/shell.zig");
+const printk = tty.printk;
 
 pub fn main() void {
     printk("hello, \x1b[32m{d}\x1b[0m\n", .{42});
```
Comment (jgiron42) [inline src/kernel.zig:6]: the motd would be better in the default/shell

File: src/shell/Shell.zig [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        const Hook: type = ?*const fn (*Self) void;
+        const ErrorHook: type = ?*const fn (*Self, err: anyerror) void;
+
+        const Hooks: type = struct {
+            pre_process: Hook = null,
+            post_process: Hook = null,
+            pre_cmd: Hook = null,
+            on_error: ErrorHook = null,
+        };
```
Comment (jgiron42) [inline src/shell/Shell.zig:26]: Ooooh

File: src/shell/Shell.zig [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        pub fn init(
+            reader: ft.io.AnyReader,
+            writer: ft.io.AnyWriter,
+            config: Config,
+            hooks: struct {
+                pre_process: ?*const anyopaque = null,
+                post_process: ?*const anyopaque = null,
+                pre_cmd: ?*const anyopaque = null,
+                on_error: ?*const anyopaque = null,
```
Comment (jgiron42) [inline src/shell/Shell.zig:42]: why can´t hooks be of type Hooks?

File: src/shell/Shell.zig [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                }
+            }
+            return CmdError.CommandNotFound;
+        }
+
+        pub fn process_line(self: *Self) void {
+            if (self.hooks.pre_process) |hook| hook(self);
+
+            // Read a line from the reader
+            var slice = self.reader.readUntilDelimiterAlloc(allocator, '\n', 4096) catch |e| {
```
Comment (jgiron42) [inline src/shell/Shell.zig:73]:
```suggestion
            var slice = self.reader.readUntilDelimiterAlloc(allocator, '\n', max_line_size) catch |e| {
```

**Review by jgiron42 — CHANGES_REQUESTED** [review]

File: src/shell/Shell.zig (hunk 4 above)
Comment (riblanc) [reply src/shell/Shell.zig:42]: Circular dependencies are a real pain... :upside_down_face:

Comment (riblanc) [reply src/shell/Shell.zig:42]: I don't remember exactly why, but I had a lot of errors about circular dependencies, and if I wanted to use the Hook type, I had to declare init as a comptime-known method (and all it args too...) or something like that, and I don't really like that

File: src/drivers/acpi/acpi.zig (hunk 1 above)
Comment (riblanc) [reply src/drivers/acpi/acpi.zig:3]: let's code a lib termcap lmao (I'm joking).

But yeah, I had the same feeling. The fact is I don't really know where I can put it... :thinking:

Comment (riblanc) [reply src/drivers/acpi/acpi.zig]: Done 27df2e6

I created a "colors" module, thus we can simply import("colors").

File: src/shell/Shell.zig (hunk 3 above)
Comment (riblanc) [reply src/shell/Shell.zig]: Yiiip yiiip
yep-derp

File: src/kernel.zig (hunk 2 above)
Comment (riblanc) [reply src/kernel.zig]: Done. 411a3d6

I added an "on_init" hook, called during initialization.

**Review by jgiron42 — APPROVED** [review]

**Pull request merged**